File: data_structures_algorithms/c/algorithms/sorting/bubble_sort.c
#include <stdio.h>

void swap(int *x, int *y) {
  int temp = *x;
  *x = *y;
  *y = temp;
}

void BubbleSort(int A[], int n) {
  int i, j;

  for (i = 0; i < n-1; i++) {
    for (j = 0; j < n - 1 - i; j++) {
      if (A[j] > A[j+1]) {
        swap(&A[j], &A[j+1]);
      }
    }
  }
}

int main() {
  int A[] = {3, 7, 10, 12, 4, 2, 11, 44, 3}, n = 9, i;

  BubbleSort(A, n)

  for (i = 0; i < n; i++) {
    printf("%d\n", A[i]);
  }

  return 0;
}

File: data_structures_algorithms/python/algorithms/searching_sorting/bubble_sort.py
""" The bubble sort makes multiple passes through a list.
It compares adjacent elements and exchanges those that are out of
order.

Each pass through the list places the next largest value in its
proper place.

Each item "bubbles" up to the location where it belongs.

Time complexity: O(n ^ 2)
Space complexity: O(1)
"""


from typing import List


def bubble_sort(arr: List[int]) -> List[int]:
    # Go backwards through the array, starting from the second to last
    # element
    for n in range(len(arr) - 1, 0, -1):
        for k in range(n):
            # If arr[k] is greater than arr[k + 1], swap
            # the values
            if arr[k] > arr[k + 1]:
                temp = arr[k]  # Temporarily store arr[k]
                arr[k] = arr[k + 1]  # Swap arr[k] with arr[k + 1]
                arr[k + 1] = temp  # Change arr[k + 1] to the value of temp

    return arr
